**Helpful Material:**

1) https://www.shanelynn.ie/summarising-aggregation-and-grouping-data-in-python-pandas/

2) Helpful function: df.value_counts()

**Learning objective**
*  EDA
*  Getting familiar with datetime operations


In [1]:
# Loading Libraries
import numpy as np
import pandas as pd

import warnings
warnings.filterwarnings("ignore")

In [2]:
chicago_data = pd.read_csv("https://raw.githubusercontent.com/dphi-official/Datasets/master/Chicago_Crime_Detective.csv")

Use the raw data github link above.

Data Columns details:
1. ID: a unique identifier for each observation.

2. Date: the date the crime occurred.
LocationDescription: the location where the crime occurred.

3. Arrest: whether or not an arrest was made for the crime (TRUE if an arrest was made, and FALSE if an arrest was not made).

4. Domestic: whether or not the crime was a domestic crime, meaning that it was committed against a family member (TRUE if it was domestic, and FALSE if it was not domestic).

5. Beat: the area, or "beat" in which the crime occurred. This is the smallest regional division defined by the Chicago police department.

6. District: the police district in which the crime occured. Each district is composed of many beats, and are defined by the Chicago Police Department.

7. CommunityArea: the community area in which the crime occurred. Since the 1920s, Chicago has been divided into what are called "community areas", of which there are now 77. The community areas were devised in an attempt to create socially homogeneous regions.

8. Year: the year in which the crime occurred.
Latitude: the latitude of the location at which the crime occurred.

In [3]:
chicago_data.columns

Index(['Unnamed: 0', 'ID', 'Date', 'LocationDescription', 'Arrest', 'Domestic',
       'Beat', 'District', 'CommunityArea', 'Year'],
      dtype='object')

In [4]:
chicago_data['formatDate'] = pd.to_datetime(chicago_data['Date'])
chicago_data['formatDate']

0        2012-12-31 23:15:00
1        2012-12-31 22:00:00
2        2012-12-31 22:00:00
3        2012-12-31 22:00:00
4        2012-12-31 21:30:00
                 ...        
191636   2001-01-01 00:05:00
191637   2001-01-01 00:01:00
191638   2001-01-01 00:01:00
191639   2001-01-01 00:01:00
191640   2001-01-01 00:01:00
Name: formatDate, Length: 191641, dtype: datetime64[ns]

In [5]:
chicago_data['month'] = pd.DatetimeIndex(chicago_data['formatDate']).month
chicago_data['month_year'] = pd.to_datetime(chicago_data['formatDate']).dt.to_period('M')
chicago_data['weekdays'] = pd.to_datetime(chicago_data['formatDate']).dt.dayofweek
chicago_data.head()

Unnamed: 0       ID            Date            LocationDescription  Arrest  \
0           0  8951354  12/31/12 23:15                         STREET   False   
1           1  8951141  12/31/12 22:00                         STREET   False   
2           2  8952745  12/31/12 22:00  RESIDENTIAL YARD (FRONT/BACK)   False   
3           3  8952223  12/31/12 22:00                         STREET   False   
4           4  8951608  12/31/12 21:30                         STREET   False   

   Domestic  Beat District CommunityArea  Year          formatDate  month  \
0     False   623        6            69  2012 2012-12-31 23:15:00     12   
1     False  1213       12            24  2012 2012-12-31 22:00:00     12   
2     False  1622       16            11  2012 2012-12-31 22:00:00     12   
3     False   724        7            67  2012 2012-12-31 22:00:00     12   
4     False   211        2            35  2012 2012-12-31 21:30:00     12   

  month_year  weekdays  
0    2012-12         0  
1    2012-12         0  
2    2012-12         0  
3    2012-12         0  
4    2012-12         0

In [6]:
chicago_data.columns

Index(['Unnamed: 0', 'ID', 'Date', 'LocationDescription', 'Arrest', 'Domestic',
       'Beat', 'District', 'CommunityArea', 'Year', 'formatDate', 'month',
       'month_year', 'weekdays'],
      dtype='object')

In [7]:
chicago_data.describe()

Unnamed: 0            ID           Beat           Year  \
count  191641.000000  1.916410e+05  191641.000000  191641.000000   
mean    95820.000000  4.968629e+06    1258.810771    2006.134940   
std     55322.135806  2.342997e+06     702.530465       3.520285   
min         0.000000  1.310022e+06     111.000000    2001.000000   
25%     47910.000000  2.832144e+06     722.000000    2003.000000   
50%     95820.000000  4.762956e+06    1121.000000    2006.000000   
75%    143730.000000  7.201878e+06    1733.000000    2009.000000   
max    191640.000000  9.181151e+06    2535.000000    2012.000000   

               month       weekdays  
count  191641.000000  191641.000000  
mean        6.611518       2.996238  
std         3.433006       1.985848  
min         1.000000       0.000000  
25%         4.000000       1.000000  
50%         7.000000       3.000000  
75%        10.000000       5.000000  
max        12.000000       6.000000

In [8]:
chicago_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 191641 entries, 0 to 191640
Data columns (total 14 columns):
Unnamed: 0             191641 non-null int64
ID                     191641 non-null int64
Date                   191641 non-null object
LocationDescription    191641 non-null object
Arrest                 191641 non-null bool
Domestic               191641 non-null bool
Beat                   191641 non-null int64
District               148589 non-null object
CommunityArea          167038 non-null object
Year                   191641 non-null int64
formatDate             191641 non-null datetime64[ns]
month                  191641 non-null int64
month_year             191641 non-null period[M]
weekdays               191641 non-null int64
dtypes: bool(2), datetime64[ns](1), int64(6), object(4), period[M](1)
memory usage: 17.9+ MB


In [9]:
chicago_data['month'].value_counts()

10    17086
7     16801
8     16572
12    16426
11    16063
9     16060
1     16047
5     16035
6     16002
3     15758
4     15280
2     13511
Name: month, dtype: int64

In [10]:
chicago_data['formatDate'].value_counts()

2008-05-16 00:00:00    11
2004-04-13 21:00:00    10
2001-10-17 22:00:00    10
2005-09-17 22:00:00    10
2001-10-26 22:00:00     9
                       ..
2011-09-05 13:00:00     1
2002-09-01 22:45:00     1
2003-12-16 21:55:00     1
2010-10-18 20:00:00     1
2004-08-31 00:20:00     1
Name: formatDate, Length: 131680, dtype: int64

In [11]:
chicago_data['Year'].value_counts()

2001    20669
2002    18753
2004    16862
2003    16657
2005    16484
2006    16098
2011    15637
2010    15497
2008    14445
2007    14280
2012    14092
2009    12167
Name: Year, dtype: int64

In [12]:
chicago_data['month_year'].value_counts()

2001-10    2291
2001-09    1972
2001-07    1952
2001-11    1869
2001-08    1795
           ... 
2009-02     920
2009-06     896
2009-04     893
2010-02     883
2009-05     861
Freq: M, Name: month_year, Length: 144, dtype: int64

In [13]:
chicago_data['weekdays'].value_counts()

4    29284
2    27416
0    27397
3    27319
5    27118
1    26791
6    26316
Name: weekdays, dtype: int64

In [14]:
median_date = pd.to_datetime(np.median(pd.to_numeric(chicago_data['formatDate'])))
median_date

Timestamp('2006-05-21 12:30:00')

**Question 1:** What is the Month and Year of the median date in our dataset? *

Enter your answer as "Month Year", without the quotes. (Ex: if the answer was 2008-03-28, you would give the answer "March 2008", without the quotes.)

**Your answer:** **May 2006**

In [15]:
chicago_data['month'].value_counts()

10    17086
7     16801
8     16572
12    16426
11    16063
9     16060
1     16047
5     16035
6     16002
3     15758
4     15280
2     13511
Name: month, dtype: int64

**Question 2:** In which month did the lowest motor vehicle thefts occur? *

(Ex: if the answer was March, you would give the answer "March", without the quotes.)

**Your answer:** **February**

**Question 3:** In which month did the lowest motor vehicle thefts occur? *
(Ex: if the answer was March, you would give the answer "March", without the quotes.)

**Your answer:** **February**


In [16]:
chicago_data['weekdays'].value_counts()

4    29284
2    27416
0    27397
3    27319
5    27118
1    26791
6    26316
Name: weekdays, dtype: int64

**Question 4:** On which weekday did most motor vehicle thefts occur? *
(Ex: if the answer was Monday, you would give the answer "Monday", without the quotes.)

**Your answer:** **Friday**


In [17]:
chicago_data.groupby('Arrest')['month'].value_counts()

Arrest  month
False   10       15744
        7        15477
        8        15243
        12       15029
        5        14848
        9        14812
        11       14807
        6        14772
        1        14612
        3        14460
        4        14028
        2        12273
True    1         1435
        12        1397
        10        1342
        8         1329
        7         1324
        3         1298
        11        1256
        4         1252
        9         1248
        2         1238
        6         1230
        5         1187
Name: month, dtype: int64

In [18]:
chicago_data['month'].value_counts()

10    17086
7     16801
8     16572
12    16426
11    16063
9     16060
1     16047
5     16035
6     16002
3     15758
4     15280
2     13511
Name: month, dtype: int64

**Question 5:** Each observation in the dataset represents a motor vehicle theft, and the Arrest variable indicates whether an arrest was later made for this theft. Which month has the largest number of motor vehicle thefts for which an arrest was made? *
(Ex: if the answer was March, you would give the answer "March", without the quotes.)

**Your answer:** **January**


In [19]:
chicago_data['Year'].value_counts()

2001    20669
2002    18753
2004    16862
2003    16657
2005    16484
2006    16098
2011    15637
2010    15497
2008    14445
2007    14280
2012    14092
2009    12167
Name: Year, dtype: int64

**Question 6:** Select the true statements *
1. In general, does it look like crime decreases from 2002 - 2012?
2. In general, does it look like crime increases from 2005 - 2008?
3. In general, does it look like crime increases from 2009 - 2011?

**Your answer:** **1. In general, does it look like crime decreases from 2002 - 2012**

In [20]:
x = chicago_data.groupby(['Arrest','month'])['month'].value_counts()
x

Arrest  month  month
False   1      1        14612
        2      2        12273
        3      3        14460
        4      4        14028
        5      5        14848
        6      6        14772
        7      7        15477
        8      8        15243
        9      9        14812
        10     10       15744
        11     11       14807
        12     12       15029
True    1      1         1435
        2      2         1238
        3      3         1298
        4      4         1252
        5      5         1187
        6      6         1230
        7      7         1324
        8      8         1329
        9      9         1248
        10     10        1342
        11     11        1256
        12     12        1397
Name: month, dtype: int64

In [21]:
x[12:18]

Arrest  month  month
True    1      1        1435
        2      2        1238
        3      3        1298
        4      4        1252
        5      5        1187
        6      6        1230
Name: month, dtype: int64

In [22]:
x[12:18].sum()

7640

In [23]:
x[18:24]

Arrest  month  month
True    7      7        1324
        8      8        1329
        9      9        1248
        10     10       1342
        11     11       1256
        12     12       1397
Name: month, dtype: int64

In [24]:
x[18:24].sum()

7896

In [25]:
y = chicago_data.groupby(['Arrest','Year'])['Year'].value_counts()
y

Arrest  Year  Year
False   2001  2001    18517
        2002  2002    16638
        2003  2003    14859
        2004  2004    15169
        2005  2005    14956
        2006  2006    14796
        2007  2007    13068
        2008  2008    13425
        2009  2009    11327
        2010  2010    14796
        2011  2011    15012
        2012  2012    13542
True    2001  2001     2152
        2002  2002     2115
        2003  2003     1798
        2004  2004     1693
        2005  2005     1528
        2006  2006     1302
        2007  2007     1212
        2008  2008     1020
        2009  2009      840
        2010  2010      701
        2011  2011      625
        2012  2012      550
Name: Year, dtype: int64

In [26]:
y[12:18]

Arrest  Year  Year
True    2001  2001    2152
        2002  2002    2115
        2003  2003    1798
        2004  2004    1693
        2005  2005    1528
        2006  2006    1302
Name: Year, dtype: int64

In [27]:
y[18:24]

Arrest  Year  Year
True    2007  2007    1212
        2008  2008    1020
        2009  2009     840
        2010  2010     701
        2011  2011     625
        2012  2012     550
Name: Year, dtype: int64

In [28]:
y[12:18].sum()

10588

In [29]:
x[18:24].sum()

7896

**Question 7:** Crimes for which more arrests were made are in the first half of the time period or the second half of the time period? *
(Note that the time period is from 2001 to 2012, so the middle of the time period is the beginning of 2007.)

1. First half
2. Second half

**Your answer:** **Second half**

In [30]:
x = chicago_data.groupby(['Arrest','Year'])['Year'].value_counts()
x[12:25]

Arrest  Year  Year
True    2001  2001    2152
        2002  2002    2115
        2003  2003    1798
        2004  2004    1693
        2005  2005    1528
        2006  2006    1302
        2007  2007    1212
        2008  2008    1020
        2009  2009     840
        2010  2010     701
        2011  2011     625
        2012  2012     550
Name: Year, dtype: int64

In [31]:
x[12:25].sum()

15536

In [32]:
temp_data = chicago_data[chicago_data['Year']==2001]
len(temp_data)

20669

In [33]:
round(temp_data['Arrest'].value_counts(normalize=True),2)

False    0.9
True     0.1
Name: Arrest, dtype: float64

**Question 8:** For what proportion of motor vehicle thefts in 2001 was an arrest made? *
**Note:** in this question and many others in the course, we are asking for an answer as a proportion. Therefore, your answer should take a value between 0 and 1

1. 0.06
2. 0.08
3. 0.10
4. 0.12

**Your answer:** **3. 0.10**

In [34]:
temp_data_a = chicago_data[chicago_data['Year']==2007]
len(temp_data_a)

14280

In [35]:
round(temp_data_a['Arrest'].value_counts(normalize=True),2)

False    0.92
True     0.08
Name: Arrest, dtype: float64

**Question 9:** For what proportion of motor vehicle thefts in 2007 was an arrest made? (round your answer to 2 decimal places)

**Your answer:** **0.08**

In [36]:
x = chicago_data['LocationDescription'].value_counts()
x[:6]

STREET                            156564
PARKING LOT/GARAGE(NON.RESID.)     14852
OTHER                               4573
ALLEY                               2308
GAS STATION                         2111
DRIVEWAY - RESIDENTIAL              1675
Name: LocationDescription, dtype: int64

**Question 10:** Which locations are the top five locations for motor vehicle thefts, excluding the "Other" category? *

1. Bank
2. Gas Station
3. Hotel/Motel
4. Street
5. Car Wash
6. Restaurant
7. Parking Lot/Garage (Non-Residential)
8. Alley
9. Driveway (Residential)
10. Vacant Lot/Land

**Your answer:**  
**4 STREET,  
7 PARKING LOT/GARAGE(NON.RESID.),  
8 ALLEY,  
2 GAS STATION,  
9 DRIVEWAY - RESIDENTIAL**

In [37]:
Top5 = chicago_data[(chicago_data['LocationDescription']== 'STREET') | (chicago_data['LocationDescription']== 'PARKING LOT/GARAGE(NON.RESID.)') | (chicago_data['LocationDescription']== 'ALLEY') | (chicago_data['LocationDescription']== 'GAS STATION') | (chicago_data['LocationDescription'] == 'DRIVEWAY - RESIDENTIAL')]
Top5.shape

(177510, 14)

In [38]:
Top0 = chicago_data[chicago_data['LocationDescription']== 'PARKING LOT/GARAGE(NON.RESID.)']
Top0.shape

(14852, 14)

In [39]:
Top1 = chicago_data[chicago_data['LocationDescription']== 'ALLEY']
Top1.shape

(2308, 14)

In [40]:
Top2 = chicago_data[chicago_data['LocationDescription']== 'GAS STATION']
Top2.shape

(2111, 14)

In [41]:
Top3 = chicago_data[chicago_data['LocationDescription']== 'DRIVEWAY - RESIDENTIAL']
Top3.shape

(1675, 14)

In [42]:
Top4 = chicago_data[chicago_data['LocationDescription']== 'STREET']
Top4.shape

(156564, 14)

**Question 11:** Create a subset of your data, only taking observations for which the theft happened in one of these five locations, and call this new data set "Top5". How many observations are in Top5? 

**Your answer:** **156564**

In [43]:
temp_data_b = chicago_data[chicago_data['LocationDescription']== 'GAS STATION']
len(temp_data_b)

2111

In [44]:
temp_data_b['weekdays'].value_counts()

5    338
6    336
4    332
3    282
0    280
2    273
1    270
Name: weekdays, dtype: int64

**Question 12:** On which day of the week do the most motor vehicle thefts at gas stations happen? *

1. Monday
2. Tuesday
3. Wednesday
4. Thursday
5. Friday
6. Saturday
7. Sunday

**Your answer:** **6. Saturday**